objectstore-service/src/backend/s3_compatible.rs

```diff
@@ -1,131 +1,117 @@
-            token_provider: Some(token_provider),
+    pub fn new(config: S3CompatibleBackendConfig) -> Self {
+        let credentials = Credentials::new(
+            config.access_key.as_deref(),
+            config.secret_key.as_deref(),
+            config.security_token.as_deref(),
+            config.session_token.as_deref(),
+            None,
+        )
+        .unwrap();
+
+        let mut bucket = Bucket::new(
+            config.bucket.as_str(),
+            Region::Custom {
+                region: config.region.clone(),
+                endpoint: match config.endpoint {
+                    Some(endpoint) => endpoint,
+                    None => format!("s3-{}.amazonaws.com", config.region),
+                },
+            },
+            credentials,
+        )
+        .unwrap();
+
+        if let Some(path_style) = config.path_style
+            && path_style
+        {
+            bucket = bucket.with_path_style();
         }
-    }
 
-    /// Formats the S3 object URL for the given key.
-    fn object_url(&self, path: &ObjectPath) -> String {
-        format!("{}/{}/{path}", self.endpoint, self.bucket)
-    }
-}
-
-impl<T> S3CompatibleBackend<T>
-where
-    T: TokenProvider,
-{
-    /// Creates a request builder with the appropriate authentication.
-    async fn request(&self, method: Method, url: impl IntoUrl) -> Result<RequestBuilder> {
-        let mut builder = self.client.request(method, url);
-        if let Some(provider) = &self.token_provider {
-            builder = builder.bearer_auth(provider.get_token().await?.as_str());
+        if let Some(request_timeout) = config.request_timeout {
+            bucket = bucket.with_request_timeout(request_timeout).unwrap();
         }
-        Ok(builder)
-    }
-
-    /// Issues a request to update the metadata for the given object.
-    async fn update_metadata(&self, path: &ObjectPath, metadata: &Metadata) -> Result<()> {
-        // NB: Meta updates require copy + REPLACE along with *all* metadata. See
-        // https://cloud.google.com/storage/docs/xml-api/put-object-copy
-        self.request(Method::PUT, self.object_url(path))
-            .await?
-            .header("x-goog-copy-source", format!("/{}/{path}", self.bucket))
-            .header("x-goog-metadata-directive", "REPLACE")
-            .headers(metadata.to_headers(GCS_CUSTOM_PREFIX, true)?)
-            .send()
-            .await?
-            .error_for_status()
-            .context("failed to update expiration time for object with TTI")?;
 
-        Ok(())
+        Self { bucket }
+    }
+
+    async fn handle_get_object_response(
+        &self,
+        _path: &ObjectPath,
+        response: ResponseData,
+    ) -> Result<Option<(Metadata, BackendStream)>> {
+        let metadata = Metadata::from_hashmap(response.headers(), "").unwrap_or_default();
+        let bytes = Bytes::from(response.to_vec());
+        let stream: BackendStream = Box::pin(stream::iter(std::iter::once(Ok(bytes))));
```

Comment:
S3 will most likely be used for large payloads, and we do not want to buffer those in memory. There should be some kind of way to get a stream from the `ResponseData`.

Reply:
There is actually: https://docs.rs/rust-s3/latest/s3/bucket/struct.Bucket.html#method.get_object_stream

But it's only compatible with async File IO type. I'm struggling to convert it into `reqwest`'s streaming format.